### PR TITLE
[BUG] The thread context is not properly cleared and messes up the traces

### PR DESCRIPTION
**Pull request by reta**
### Description
The thread context stashing messes up the propagation of the current span across thread boundaries (and even same thread boundary). It leads to split brain situation when some state is stored in `ThreadContext` and in the thread local scope, which is difficult (if possible at all) to reconcile. 

### Related Issues

Closes https://github.com/opensearch-project/OpenSearch/issues/10789

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/29040/

https://github.com/opensearch-project/OpenSearch/issues/10785

**Comment by Gaganjuneja** [comment]
Looking into it.

File: server/src/test/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorageTests.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    public void testSpanInScopeChangesThreadContext() {
+        final Span span = tracer.startSpan(SpanCreationContext.internal().name("test"));
+
+        try (SpanScope scope = tracer.withSpanInScope(span)) {
+            assertThat(threadContext.getTransient(ThreadContextBasedTracerContextStorage.CURRENT_SPAN), is(not(nullValue())));
+            assertThat(threadContextStorage.get(ThreadContextBasedTracerContextStorage.CURRENT_SPAN), is(span));
+        }
+
+        assertThat(threadContext.getTransient(ThreadContextBasedTracerContextStorage.CURRENT_SPAN), is(not(nullValue())));
```
[the remark this answers is not in the document]
Comment (reta) [reply server/src/test/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorageTests.java:111]: The `SpanReference` is not accessible here, this class is package private, the next assertion does this check without accessing `SpanReference` directly.

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/SpanReference.java [hunk 2]
```diff
@@ -17,7 +17,6 @@
  */
 @InternalApi
 final class SpanReference {
-
```
[the remark this answers is not in the document]
Comment (reta) [reply libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/SpanReference.java]: Reverted, thank you

**Review by Gaganjuneja — APPROVED** [review]

**Comment by reta** [comment]
@andrross may I ask you please to approve, we worked with @Gaganjuneja to confirm the "mess" it gone for now but the work will continue, thank you.

File: CHANGELOG.md [hunk 3]
```diff
@@ -90,6 +90,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Fix 'org.apache.hc.core5.http.ParseException: Invalid protocol version' under JDK 16+ ([#4827](https://github.com/opensearch-project/OpenSearch/pull/4827))
 - Fix compression support for h2c protocol ([#4944](https://github.com/opensearch-project/OpenSearch/pull/4944))
 - Don't over-allocate in HeapBufferedAsyncEntityConsumer in order to consume the response ([#9993](https://github.com/opensearch-project/OpenSearch/pull/9993))
+- [BUG] The thread context is not properly cleared and messes up the traces ([#10873](https://github.com/opensearch-project/OpenSearch/pull/10873))
```
Comment (dblock) [inline CHANGELOG.md:93]: This should say `Fix ....` like the others.

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultSpanScope.java [hunk 4]
```diff
@@ -44,23 +44,23 @@ private DefaultSpanScope(Span span, SpanScope previousSpanScope, TracerContextSt
     public static SpanScope create(Span span, TracerContextStorage<String, Span> tracerContextStorage) {
         final SpanScope beforeSpanScope = spanScopeThreadLocal.get();
         SpanScope newSpanScope = new DefaultSpanScope(span, beforeSpanScope, tracerContextStorage);
-        spanScopeThreadLocal.set(newSpanScope);
```
Comment (dblock) [inline libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultSpanScope.java:47]: So create no longer attaches?

**Review by dblock — APPROVED** [review]
There should be a unit test for `DefaultSpanScope` that ensures attach/detach works in the various scenarios (e.g. exception).

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultSpanScope.java (hunk 4 above)
Comment (reta) [reply libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultSpanScope.java:47]: Nope and it shouldn't, the `attach` does

**Comment by reta** [comment]
> There should be a unit test for `DefaultSpanScope` that ensures attach/detach works in the various scenarios (e.g. exception).

We do have tests for that actually, the reason why everything still work is that the `DefaultSpanScope` was always created manipulated this way: `DefaultSpanScope.create(...).attach()` so functionally there are no changes in this regards.

**Review by dblock — APPROVED** [review]

**Pull request merged**